### ISSUE TITLE
`id` command cannot show correct group information inside container with mounted /etc/group and /etc/passwd files 

### ISSUE THREAD

**Issue by cookieisaac** [moby/moby#26840]
**Description**

`id` command cannot show correct user information with mounted `/etc/passwd` and `/etc/group` file for docker ubuntu 1.12.1 and suse 1.11.1 distribution

**Steps to reproduce the issue:**
1 Prepare the following file called `passwd` with the following content

```
shuffle:x:5563:2222::/home/shuffle:
```

2  Prepare the following file called `group` with the following content

```
spark:x:2222
```

3  Start a docker container with mounted `/etc/passwd` and `/etc/group` files

```
docker run -it -v `pwd`/group:/etc/group -v `pwd`/passwd:/etc/passwd -u 5563 --rm ubuntu:latest /bin/bash
```

4  Inside the container, issue `id` and `id shuffle` command separately

**Describe the results you received:**
Inside the container, here's my test result

```
root@my-machine:/tmp/test# docker run -it -v `pwd`/group:/etc/group -v `pwd`/passwd:/etc/passwd -u 5563 --rm ubuntu:latest /bin/bash
groups: cannot find name for group ID 0
shuffle@6b8a345d14ec:/$ id
uid=5563(shuffle) gid=0 groups=2222(spark)   [<===== Note the gid is 0, instead of 2222]
shuffle@6b8a345d14ec:/$ id shuffle
uid=5563(shuffle) gid=2222(spark) groups=2222(spark) [<===== Note that gid is correctly resolved for user shuffle]
shuffle@6b8a345d14ec:/$ cat /etc/group
spark:x:2222
shuffle@6b8a345d14ec:/$ cat /etc/passwd
shuffle:x:5563:2222::/home/shuffle:
```

**Describe the results you expected:**

`id` and `id shuffle` should both render exact same result `uid=5563(shuffle) gid=2222(spark) groups=2222(spark)`

**Additional information you deem important (e.g. issue happens only occasionally):**

The case cannot be reproduced on RHEL7.1/7.2  with docker 1.10.3, and Ubuntu 16.04 with docker 1.10.3

The case is reproducible on RHEL7.1/7.2 with docker 1.12.1, Ubuntu 16.04 with docker 1.12.1, SUSE12 with docker 1.11.2

**Output of `docker version`:**

```
# docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
# docker info
Containers: 3
 Running: 2
 Paused: 0
 Stopped: 1
Images: 6
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 12
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-34-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859 GiB
Name: asc-13
ID: SQ3P:KZMH:R37D:E5ZH:KOW3:YT4E:T5BZ:FYUK:KNIS:YESD:Y7JW:DOXS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Also reproduced in Operating System: `SUSE Linux Enterprise Server 12` with Kernel Version: `3.12.28-4-default` with Docker Version: `1.11.2`

**Comment by cpuguy83** [comment]
ping @crosbymichael 

I imagine this is because you specified `-u 5563` without specifying a group ID.
So the process is actually started with group 0 even though /etc/group has the user in an additional group.

But I'm not sure what the "correct" thing to do is here.

**Comment by cpuguy83** [comment]
btw, if you want to specify custom groups, it would be better to do `-u 5563:2222 --group-add <gid> --group-add <gid>`

**Comment by cookieisaac** [comment]
@cpuguy83 Thanks for the information, that actually helped me work around my current problem.

[Note that I only used `-u 5563` because docker doesn't allow me to use `-u shuffle` anymore... I thought previous behavior is that when I use string name `shuffle`, it would use all the information in `/etc/passwd` including the primary group information]

**Comment by cpuguy83** [comment]
@cookieisaac You could just add the user with the uid/gid you want as part of the image.

**Comment by cookieisaac** [comment]
@cpuguy83 In the real use case, I cannot assume the existence of any user builtin the image, so I have to prepare a /etc/passwd and mount it during runtime.

**Comment by cpuguy83** [comment]
@cookieisaac Why, these are your images, no?

**Comment by cookieisaac** [comment]
@cpuguy83 Nope. I have a python application that takes images from user, and start container based on user requirements, i.e., start the container with host user `shuffle` account. My application handles all the impersonation process (such as preparing /etc/group, /etc/passwd) without prior knowledge of any user-provided images.

**Comment by frastel** [comment]
I have a similar problem. I am on a CoreOS machine and do a `docker run` command with my `core` user which I am used from within older CoreOS machines. `/etc/group` and `/etc/passwd` are mounted, in addition the docker socket too. The whole command is similar to this [docker run wrapper](https://github.com/chefkoch-dev/docker-run/blob/master/run).

The problem:
The core user within the `docker run` command has broken groups and thus he couldn't access the mounted docker socket. The socket is only reachable by the `docker` group. Accessing it via root does work however.

It works with: 
* CoreOS machines with Docker 1.9.1 installed

It does not work with:
* CoreOS machine, Release 1192.2.0, Docker 1.12.1 (detailed system information below)
* CoreOS machine, Release 1153.4.0, Docker 1.11.2

My system:

Host system where Docker is installed
`CoreOS beta (1192.2.0)`

Guest system for the docker run command
```
core@e9e5be94c055:/code$ cat  /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

Docker Version
```
core@e9e5be94c055:/code$ docker version
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   7a86f89
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   7a86f89
 Built:        
 OS/Arch:      linux/amd64
```

The workaround with `--group-add ` works for me currently but I think there will occure sideeffects sooner or later.

**Comment by fantapop** [comment]
I'm running into this as well and having to use the workaround.

**Comment by thaJeztah** [comment]
Let me close this ticket as it went stale

I think this is largely expected when only specifying a user-ID (but no group); the given user may be looked up inside the container's filesystem when the container is created, but that does not include files mounted in. So lookup won't have information from a custom `/etc/password` or `/etc/group` at that time.
